homeassistant/components/websocket_api/commands.py

```diff
@@ -310,6 +319,69 @@ async def handle_integration_setup_info(
     )
 
 
+@decorators.websocket_command({vol.Required("type"): "integration/log_info"})
+@decorators.async_response
+async def handle_integration_log_info(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle integrations logger info."""
+    connection.send_result(
+        msg["id"],
+        [
+            {
+                "domain": integration,
+                "level": logging.getLogger(
+                    f"homeassistant.components.{integration}"
+                ).getEffectiveLevel(),
+            }
+            for integration in async_get_loaded_integrations(hass)
+        ],
+    )
+
+
+@decorators.websocket_command(
+    {
+        vol.Required("type"): "integration/log_level",
+        vol.Required("integration"): str,
+        vol.Required("level"): vol.In(
+            [
+                "CRITICAL",
+                "FATAL",
+                "ERROR",
+                "WARNING",
+                "WARN",
+                "INFO",
+                "DEBUG",
+                "NOTSET",
+            ]
+        ),
+    }
+)
+@decorators.async_response
+async def handle_integration_log_level(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle setting integration log level."""
+    try:
+        integration = await async_get_integration(hass, msg["integration"])
+    except IntegrationNotFound:
+        connection.send_error(msg["id"], const.ERR_NOT_FOUND, "Integration not found")
+        return
+    loggers = [f"homeassistant.components.{msg['integration']}"]
+    if integration.loggers:
+        loggers.extend(integration.loggers)
+    await _handle_call_service(
```

Comment:
Our websocket integration allows creating subscriptions. See how ZHA does it for their logging forwarding:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/api.py#L207-L221

we can do something similar but instead just attach temporary listeners to these logging objects.

Reply:
Then the frontend gets all the log lines, can show them as they see fit, and allow user to download it.

Reply:
And by allowing the user to see the debug logs, we don't have to teach users about terminals + they know what they are capturing (kinda)

Reply:
I've changed my mind about having a ZHA-like API. Using the normal logging is good, as we can always decide to stream that log to the frontend and so capture all use cases.